fabric-tag-extensions-v0/src/main/java/net/fabricmc/fabric/api/tag/TagFactory.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.tag;
+
+import java.util.function.Supplier;
+
+import net.minecraft.block.Block;
+import net.minecraft.entity.EntityType;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.item.Item;
+import net.minecraft.tag.BlockTags;
+import net.minecraft.tag.EntityTypeTags;
+import net.minecraft.tag.FluidTags;
+import net.minecraft.tag.GameEventTags;
+import net.minecraft.tag.ItemTags;
+import net.minecraft.tag.Tag;
+import net.minecraft.tag.TagGroup;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.event.GameEvent;
+
+import net.fabricmc.fabric.impl.tag.extension.TagFactoryImpl;
+
+/**
+ * A factory for accessing datapack tags.
+ */
+public interface TagFactory<T> {
+	TagFactory<Item> ITEM = of(ItemTags::getTagGroup);
+	TagFactory<Block> BLOCK = of(BlockTags::getTagGroup);
+	TagFactory<Fluid> FLUID = of(FluidTags::getTagGroup);
+	TagFactory<GameEvent> GAME_EVENT = of(GameEventTags::getTagGroup);
+	TagFactory<EntityType<?>> ENTITY_TYPE = of(EntityTypeTags::getTagGroup);
+	TagFactory<Biome> BIOME = of(Registry.BIOME_KEY, "tags/biomes");
```

Comment:
Why biomes specifically?

Reply:
It was my suggestion as biomes tags can be very useful and should be encouraged to be used in the appropriate situations.

Reply:
It should only be created as needed. That way, no registries other than vanilla ones get special treatment.

Reply:
One of the big issue with those non-built-in-tags factories are the path is basically up to the modder, and you cannot have multiple paths.
Hopefully all modders will use the standardized paths, but you can't really guarantee it.
I expect the biomes tags to be widely used in the future so I don't see why not providing it as "built-in". Plus it would encourage the use of those.

Reply:
Biome tags specifically are also something I've seen requested as a feature and asked about a lot on Discord, so it's good to have them in Fabric API.

Reply:
Surely a modder can use `of(Registry.BIOME_KEY, "tags/biomes")` themselves if they need biome tags. It could even have a simplified overload of just `of(Registry.BIOME_KEY)` to prevent mismatches in the path.

Reply:
Looked again at Registry keys, and turns out that not exposing the path in API and creating it from the registry key would be much more safer and avoid modders having different registry base tag paths.

Reply:
That does sort of run into problems if people have tags for modded registries - "should a non-minecraft namespace be somehow included to prevent conflicts?" etc

Reply:
Base tag paths should have those rules:
 - starts with `tags/`
 - if the `registry` field of the `RegistryKey` is not `Registry.ROOT` then append it, if the namespace is not minecraft then append it too as `namespace/path`.
 - append the value field, if the namespace is not minecraft then append it too as `namespace/path`
 - ends with `s`

Reply:
I don't think the `s` requirement will be possible due to exceptions. For example, the `fabric:cactus` registry would have a path of `tags/fabric/cactuss`.